src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.CompilationTrackerState.cs

```diff
@@ -39,21 +41,52 @@ private partial class CompilationTracker
                 /// </summary>
                 public readonly bool DocumentsAreFinal;
 
+                /// <summary>
+                /// Whether the generated documents are frozen and generators should never be ran again, ever, even if a document
+                /// is later changed. This is used to ensure that when we produce a frozen solution for partial semantics,
+                /// further downstream forking of that solution won't rerun generators. This is because of two reasons:
+                /// <list type="number">
+                /// <item>Generally once we've produced a frozen solution with partial semantics, we now want speed rather
+                /// than accuracy; a generator running in a later path will still cause issues there.</item>
+                /// <item>The frozen solution with partial semantics makes no guarantee that other syntax trees exist or
+                /// whether we even have references -- it's pretty likely that running a generator might produce worse results
+                /// than what we originally had.</item>
+                /// </list>
+                /// </summary>
+                public readonly bool DocumentsAreFinalAndFrozen;
+
                 public CompilationTrackerGeneratorInfo(
                     TextDocumentStates<SourceGeneratedDocumentState> documents,
                     GeneratorDriver? driver,
-                    bool documentsAreFinal)
+                    bool documentsAreFinal,
+                    bool documentsAreFinalAndFrozen = false)
                 {
                     Documents = documents;
                     Driver = driver;
                     DocumentsAreFinal = documentsAreFinal;
+                    DocumentsAreFinalAndFrozen = documentsAreFinalAndFrozen;
+
+                    // If we're frozen, that implies final as well
+                    Contract.ThrowIfTrue(documentsAreFinalAndFrozen && !documentsAreFinal);
                 }
 
                 public CompilationTrackerGeneratorInfo WithDocumentsAreFinal(bool documentsAreFinal)
-                    => DocumentsAreFinal == documentsAreFinal ? this : new(Documents, Driver, documentsAreFinal);
+                {
+                    // If we're already frozen, then we don't allow any further changes to whether we need to be regenerated;
+                    // otherwise we can return the same copy if nothing changed.
+                    if (DocumentsAreFinalAndFrozen || DocumentsAreFinal == documentsAreFinal)
```

Comment:
@CyrusNajmabadi: not quite, this is the key here: some places say WithDocumentsAreFinal(true) when we have ran generators, namely when we're finalizing. We also have places that say WithDocumentsAreFinal(false), which is when we are forking because we know at that point the generated results are stale. The idea here is once we're FinalAndFrozen, we're disregarding the WithDocumentsAreFinal(false). That is admittedly a bit tricky (since we're ignoring the request of the caller), but the hope is prevents everybody else from having to add a "oh, this is already frozen so never mind". I'm not sure there are many callers who do WithDocumentsAreFinal(false) but the hope is to avoid writing future bugs that way.

Reply:
> We also have places that say WithDocumentsAreFinal(false),

ohhhhhhhhhhhhhhhhhh..    please doc this code to make that clear (for weak brained people like me).

Reply:
I'll beef up the comment.